tests/system_tests/backend_fixtures.py

```diff
@@ -0,0 +1,213 @@
+from __future__ import annotations
+
+import json
+import secrets
+import subprocess
+import sys
+from collections import deque
+from dataclasses import InitVar, asdict, dataclass, field
+from pathlib import Path
+from string import ascii_lowercase, digits
+from typing import Any, ClassVar, Final, Generator, Literal
+
+import httpx
+import pytest
+
+if sys.version_info < (3, 12):
+    from typing_extensions import dataclass_transform
+else:
+    from typing import dataclass_transform
+
+
+#: The root directory of this repo.
+REPO_ROOT: Final[Path] = Path(__file__).parent.parent.parent
+
+
+@dataclass(frozen=True)
+class LocalWandbBackendAddress:
+    host: str
+    base_port: int
+    fixture_port: int
+
+    @property
+    def base_url(self) -> str:
+        return f"http://{self.host}:{self.base_port}"
+
+    @property
+    def fixture_service_url(self) -> str:
+        return f"http://{self.host}:{self.fixture_port}"
+
+
+@pytest.fixture(scope="session")
+def local_wandb_backend() -> LocalWandbBackendAddress:
+    """Fixture that starts up or connects to the local-testcontainer.
+
+    This does not patch WANDB_BASE_URL! Use `use_local_wandb_backend` instead.
+    """
+    return _local_wandb_backend(name="wandb-local-testcontainer")
+
+
+@pytest.fixture(scope="session")
+def local_wandb_backend_importers() -> LocalWandbBackendAddress:
+    """Fixture that starts up or connects to a second local-testcontainer.
+
+    This is used by importer tests, to move data between two backends.
+    """
+    return _local_wandb_backend(name="wandb-local-testcontainer-importers")
+
+
+def _local_wandb_backend(name: str) -> LocalWandbBackendAddress:
+    tool_file = REPO_ROOT / "tools" / "local_wandb_server.py"
+
+    result = subprocess.run(
+        [
+            "python",
+            tool_file,
+            "connect",
+            f"--name={name}",
+        ],
+        stdout=subprocess.PIPE,
+    )
+
+    if result.returncode != 0:
+        raise AssertionError(
+            "`python tools/local_wandb_server.py connect` failed. See stderr."
+            " Did you run `python tools/local_wandb_server.py start`?"
+        )
+
+    output = json.loads(result.stdout)
+    address = LocalWandbBackendAddress(
+        host="localhost",
+        base_port=int(output["base_port"]),
+        fixture_port=int(output["fixture_port"]),
+    )
+    return address
+
+
+@pytest.fixture(scope="function")
+def use_local_wandb_backend(
+    local_wandb_backend: LocalWandbBackendAddress,
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    """Fixture that patches WANDB_BASE_URL to point to the local container.
+
+    Unlike `local_wandb_backend`, this is function-scoped, so cannot be used
+    in session-scoped fixtures.
+    """
+    monkeypatch.setenv("WANDB_BASE_URL", local_wandb_backend.base_url)
```

Comment:
**note:** Code **above** here was copied over from `test/system_tests/conftest.py` and should be entirely (or almost entirely) unchanged